### PR TITLE
feat: user roles edit APIs

### PR DESCRIPTION
**Pull request by shadinaif**
feat: user roles edit APIs

* add user roles: done
* delete roles for user: done
* update roles for user: done

### CODE REVIEW

**Comment by OmarIthawi** [comment]
Amazing @shadinaif! Let me know if there are specific parts I should be focused on.

File: futurex_openedx_extensions/helpers/extractors.py [hunk 1]
```diff
@@ -2,27 +2,133 @@
...
+    :type fields: List[str]
+    :param separator: Separator to use between field values.
+    :type separator: str
+    :param replace_none: String to replace None values with.
+    :type replace_none: str
+    :return: Hashcode string.
+    """
+    return separator.join(replace_none if value_dict[field] is None else str(value_dict[field]) for field in fields)
+
+
```
Comment (OmarIthawi) [inline futurex_openedx_extensions/helpers/extractors.py:105]:
```suggestion
from dataclasses import dataclass


@dataclass
class Hashcode:
    """Class for keeping track of an item in inventory."""
    role: str
    org: str
    course_id: str 

    def __has__(self) -> str:
        return '{self.role},{self.org},{self.course}'


```

File: futurex_openedx_extensions/helpers/extractors.py [hunk 2]
```diff
@@ -2,27 +2,133 @@
...
+    :type value_dict: Dict[str, Any]
+    :param fields: List of field names to include in the hashcode.
+    :type fields: List[str]
+    :param separator: Separator to use between field values.
+    :type separator: str
+    :param replace_none: String to replace None values with.
+    :type replace_none: str
+    :return: Hashcode string.
+    """
+    return separator.join(replace_none if value_dict[field] is None else str(value_dict[field]) for field in fields)
```
Comment (OmarIthawi) [inline futurex_openedx_extensions/helpers/extractors.py:103]: This way the returned code can both be used as a Dict key _and_ as an information holder without needing to parse strings.

```suggestion
    return Hashcode(role, org, course_id)
```

File: futurex_openedx_extensions/helpers/roles.py [hunk 3]
```diff
@@ -628,4 +636,403 @@ def get_course_access_roles_queryset(  # pylint: disable=too-many-arguments
...
+    for hashcode in redundant_hashes:
+        try:
+            course_id: str | object = hashcode.split(separator)[2]
+            if course_id == 'None':
+                course_id = CourseKeyField.Empty
+
+            delete_count, _ = CourseAccessRole.objects.filter(
+                user=user,
+                role=hashcode.split(separator)[0],
+                org=hashcode.split(separator)[1],
```
Comment (OmarIthawi) [inline futurex_openedx_extensions/helpers/roles.py:694]: Now the hashcode is both a dict key and an information holder i.e. regular class.

```suggestion
                role=hashcode.role,
                org=hashcode.org,
```

**Review by OmarIthawi — COMMENTED** [review]
@shadinaif thank you. 

I'm having a note on what I think is anti-pattern with hashcode.

I don't think we should parse hashcodes after it's been made into a string.

The recommendation is to make use of `__hash__` overriding so hashcodes can be used without parsing.

File: futurex_openedx_extensions/helpers/extractors.py (hunk 2 above)
Comment (shadinaif) [reply futurex_openedx_extensions/helpers/extractors.py]: Done @OmarIthawi , please check

**Review by OmarIthawi — APPROVED** [review]
Thanks @shadinaif! It's much better now without parsing.

**Pull request merged**